### PR TITLE
Initial changes for llvm shared library build using explicit visibility annotations

### PR DESCRIPTION
**Pull request by fsfod**
These are my initial build and code changes to supporting building llvm as shared library/DLL on windows(without force exporting all symbols) and making symbol visibility hidden by default on Linux which adding explicit symbol visibility macros to the whole llvm codebase.

Updated cmake code to allow building llvm-shlib on windows by appending /WHOLEARCHIVE:lib to the linker options.
Remove the hardcoded CMake error from using LLVM_BUILD_LLVM_DYLIB on windows.
Updated CMake to define new macros to control conditional export macros in llvm/Support/Compiler.h
Use /Zc:dllexportInlines- when compiling with clang-cl on windows with a opt out CMake option to disable using it.
Replace some use of LLVM_EXTERNAL_VISIBILITY with new export macros.

Some of the cmake and code changes are based on @tstellar's earlier PR https://github.com/llvm/llvm-project/pull/67502. 

I have Windows building using clang-cl, while for MSVC its at-least able to build libllvm, but some tests can't build because llvm iterator template metaprogramming that doesn't work well with dllexport. Linux should build without issue. My full branch is here https://github.com/fsfod/llvm-project/tree/llvm-export-api-20.0 and including all the auto generated export macros from clang tooling based system.

### CODE REVIEW

File: llvm/lib/Support/BLAKE3/CMakeLists.txt
Comment (tstellar) [inline llvm/lib/Support/BLAKE3/CMakeLists.txt:85]: Do we know why it doesn't use llvm_add_library?

Comment (fsfod) [reply llvm/lib/Support/BLAKE3/CMakeLists.txt:85]: My best guess is because it a bunch of C code and assembly files that need custom file options

File: llvm/include/llvm/CodeGen/SelectionDAGNodes.h
Comment (fsfod) [inline llvm/include/llvm/CodeGen/SelectionDAGNodes.h:1083]: I also found SystemZAddressingMode was calling one these functions outside of a debug build, but the only use was indirectly in a LLVM_DEBUG macro.

File: llvm/include/llvm/Support/Compiler.h
Comment (fsfod) [inline llvm/include/llvm/Support/Compiler.h:171]: This mode is needed for a few static libraries like tablegen don't link to the llvm shared library and if it was left exporting symbols there might be duplicate symbols errors from two different llvm static libraries linked in to an executable.

**Comment by vgvassilev** [comment]
@rnk, can you take a look? This pull request is in the context of https://llvm.org/OpenProjects.html#clang-plugins-windows and google summer of code.

File: llvm/CMakeLists.txt
Comment (rnk) [inline llvm/CMakeLists.txt:833]: Typo? INLNES vs INLINES

File: llvm/include/llvm/Support/Compiler.h
Comment (rnk) [reply llvm/include/llvm/Support/Compiler.h:171]: Can you add a comment to that effect here after the `#else`?

File: llvm/tools/llvm-shlib/CMakeLists.txt
Comment (rnk) [inline llvm/tools/llvm-shlib/CMakeLists.txt:52]: It would be good to handle this before landing. If we can't use the cmake feature, file a bug in their tracker and link to it from the comments.

File: llvm/tools/llvm-shlib/CMakeLists.txt
Comment (rnk) [inline llvm/tools/llvm-shlib/CMakeLists.txt:68]: 👍

**Review by rnk — COMMENTED** [review]
Broadly speaking, yeah, I think this is the right direction for building libLLVM.dll. I'm looking forward to it coming together. :)

I didn't hit approve since only early feedback was requested, and I think there's stuff to iterate on.

File: llvm/include/llvm/Support/Compiler.h
Comment (compnerd) [reply llvm/include/llvm/Support/Compiler.h:171]: Can we use the following pattern instead?

```c
#if defined(LLVM_BUILD_STATIC)
# define LLVM_ABI
# define LLVM_TEMPLATE_ABI
# define LLVM_EXPORT_TEMPLATE
#else
# if defined(LLVM_EXPORTS)
#   define LLVM_ABI __declspec(dllexport)
#   define LLVM_TEMPLATE_ABI
#   define LLVM_EXPORT_TEMPLATE __declspec(dllexport)
# else
#   define LLVM_ABI __declspec(dllimport)
#   define LLVM_TEMPLATE_ABI __declspec(dllimport)
#   define LLVM_EXPORT_TEMPLATE
# endif
#endif
```

If you are statically linking, you define `LLVM_BUILD_STATIC`, otherwise, if you are building the DLL, you define `LLVM_EXPORTS`.

Comment (rnk) [reply llvm/include/llvm/Support/Compiler.h:171]: The purpose of doing it that way is to make it so normal users don't have to define macros to get the right settings. But, those dllimport settings are incorrect if you're linking against LLVM statically. It raises the question of, in the future, which will be more common, using LLVM.dll, or static linking? I DLLs are better for distribution, so we should lean into and go with @compnerd's suggested direction.

Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h:171]: i wasn't sure if @compnerd meant setting LLVM_BUILD_STATIC for specific libraries like tablegen when building llvm as a shared library.

Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h:171]: There is also the weird edge case of statically linking and LLVM_EXPORT_SYMBOLS_FOR_PLUGINS where component symbols are exported from the tool executables.

Comment (compnerd) [reply llvm/include/llvm/Support/Compiler.h:171]: @rnk exactly right :) Most users wouldn't need to figure out defines, if they want static linking they specify `/DLLVM_BUILD_STATIC` (we can rename that to `/DLLVM_STATIC` or `/DLLVM_STATIC_BUILD` or something along those lines). Otherwise, they get the default behaviour of DLL linkage.

**Comment by vgvassilev** [comment]
@fsfod, @compnerd, @rnk, what is the preferred way to move forward here? If we want atomic history we should probably land the whole pull request. It is not clear to me if we can chunk it at a reasonable level...

**Comment by fsfod** [comment]
> @fsfod, @compnerd, @tstellar, @rnk, what is the preferred way to move forward here? If we want atomic history we should probably land the whole pull request. It is not clear to me if we can chunk it at a reasonable level...

The thing a lot of these changes can't go in without the rest of changes in my full branch that annotates the whole llvm codebase headers with export macros and the many fixes required from MSVC instantiating all class members. People probably still want the old windows shared library build system to also keep working before all my changes are merged so the export macros on windows would initially have to be a nop.

What i was thinking was just get the export macros definitions and cmake machinery to drive them merged first.

**Comment by compnerd** [comment]
Yeah, I think that annotating the libraries with the ABI boundaries should be good as a first step. It would allow us to split up the patch along maintainer boundaries as well hopefully and keep the size of the review more manageable.

File: llvm/include/llvm/Support/Compiler.h
Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h]: @compnerd i had push some changes to reorganize the macros

**Comment by fsfod** [comment]
I've added cmake config option(LLVM_BUILD_LLVM_DYLIB_VIS) to test explicit visibility macros on windows and Linux. 
When its off the macros are nop on windows and for Linux there is no difference with it off , except the default symbol visibility is not set to hidden for llvm components.
If option is enabled llvm shared library should build on both windows(only clang-cl this early on) and Linux with it on, but tools won't link successfully yet because only limited number of symbols are exported.

File: llvm/CMakeLists.txt
Comment (compnerd) [inline llvm/CMakeLists.txt:831]: Can we make the `option`s unconditional? The bit that I imagine that we care about is the default value which we can compute appropriately for the environment where we were previously adding the `option`.

File: llvm/include/llvm/Support/Compiler.h
Comment (compnerd) [inline llvm/include/llvm/Support/Compiler.h:118]: Clang should support this under the GNU standard. Can you add a note on why this condition is needed?

Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h:118]: I'm not sure what the need for it was, it was part of @tstellar initial changes to the file that i based my code on. I can just remove it if there wasn't some clear purpose for it.

File: llvm/CMakeLists.txt
Comment (fsfod) [reply llvm/CMakeLists.txt:831]: I had made it unconditioned before, but i reverted it back to the original code to discourage people trying to use it and reporting bugs while it won't work correctly yet. I think that was part of the intent of earlier code being hardcoded to off earlier in the file https://github.com/llvm/llvm-project/blob/d286efeb1d0d3e135901dd07535cc68aed39a758/llvm/CMakeLists.txt#L844-L860

File: llvm/include/llvm/Support/Compiler.h
Comment (tstellar) [reply llvm/include/llvm/Support/Compiler.h:118]: Clang supports it, but IIRC the rules about where the function declaration the c++11 style attributes can be placed is different between gcc in clang.  See the comments here:

https://github.com/tstellar/template-visibility/blob/testing/templates.h

Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h:118]: That file is my pain. I did add a option to the clang tool to switch between the two places where to generate the macro for for functions. I also did try generating export macros on template classes, but it just creates ton a more code that needs to be fixed for MSVC and can be problematic for class static fields when classes with instantiations outside the llvm shared library.

**Comment by vgvassilev** [comment]
@compnerd, @tstellar, @rnk, can you take a look. It seems this PR is ready to review.

File: llvm/tools/llvm-shlib/CMakeLists.txt
Comment (fsfod) [reply llvm/tools/llvm-shlib/CMakeLists.txt:55]: My guess is it could be from weird interaction from object file targets llvm uses in cmake to build both static and shared libraries.

Comment (fsfod) [reply llvm/tools/llvm-shlib/CMakeLists.txt:55]: this was the error 
```
1> [CMake] CMake Error at cmake/modules/AddLLVM.cmake:642 (add_library):
1> [CMake]   Impossible to link target 'LLVM' because the link item 'LLVMSupport',
1> [CMake]   specified without any feature or 'DEFAULT' feature, has already occurred
1> [CMake]   with the feature 'WHOLE_ARCHIVE', which is not allowed.
```

**Comment by vgvassilev** [comment]
> @compnerd, @tstellar, @rnk, can you take a look. It seems this PR is ready to review.

ping.

**Comment by compnerd** [comment]
I think that @petrhosek might have opinions on the CMake changes.

**Review by compnerd — COMMENTED** [review]
Most of the non-CMake changes seem reasonable to me.

File: llvm/include/llvm/Support/Compiler.h
Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h:178]: The idea for LLVM_ABI_IMPORT was that the per module export header could just forward to LLVM_ABI_EXPORT\LLVM_ABI_IMPORT and not repeat the logic for per compiler #ifdefs.

Comment (compnerd) [reply llvm/include/llvm/Support/Compiler.h:178]: Sorry, I'm not sure I understand. The `LLVM_ABI_IMPORT` doesn't really make sense, IMO. Consider the following:

```c
void ABI __fastcall function(void);
```

This function declaration should be annotated as `__declspec(dllexport)` when being *defined*. However, when declared as an external symbol, the _only_ difference would be that it would be annotated `__declspec(dllimport)`. The rest of the signature should be untouched (including the calling convention). You do not want to have your own declaration as that allows for it to go out of sync. So the definition of `ABI` changes between the build and use.

**Review by tstellar — COMMENTED** [review]
I don't understand some of these things as well as I used to, but I don't see any major issues with this patch.

File: llvm/include/llvm/Support/Compiler.h
Comment (compnerd) [inline llvm/include/llvm/Support/Compiler.h:201]: We really should not conflate the LLVM ABI and the LLVM C ABI.

File: llvm/include/llvm/Support/Compiler.h
Comment (compnerd) [inline llvm/include/llvm/Support/Compiler.h:168]: We should hoist this into CMake and map this to `LLVM_BUILD_STATIC`.

File: llvm/include/llvm/Support/Compiler.h
Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h:201]: The idea was it could maybe made separate later on for some kind of optimized llvmc only build. There also exported functions in Target libs like LLVMInitialize* that are used both in C and c++ API.

Comment (compnerd) [reply llvm/include/llvm/Support/Compiler.h:201]: That's actually more towards what my point was :) We shouldn't be defining the C ABI in terms of the C++ ABI. The C ABI is guaranteed stable while the C++ is not. It may be someone wants to not expose the C++ ABI (builds statically) but still build the C interface as a DSO.

File: llvm/include/llvm/Support/Compiler.h
Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h:168]: I just wanted avoid doing it in CMake yet because its more complex to do and easy to get wrong.

File: llvm/CMakeLists.txt
Comment (petrhosek) [reply llvm/CMakeLists.txt:831]: I'd use https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html.

File: llvm/include/llvm/Support/Compiler.h
Comment (compnerd) [reply llvm/include/llvm/Support/Compiler.h:168]: Okay; I guess getting it in is a good first step. The CMake bits, I think you should wait to get sign off from @petrhosek as well.

File: llvm/CMakeLists.txt
Comment (fsfod) [reply llvm/CMakeLists.txt:831]: cmake_dependent_option seems to require all conditions be true while i need an OR, the Condition Syntax note on that page seems to imply i could put parentheses around an OR expression, but the minimum version for it is higher than LLVM's.

Comment (fsfod) [reply llvm/CMakeLists.txt]: @petrhosek I pushed some changes to try and use it if you want to take a look.

**Comment by vgvassilev** [comment]
@fsfod, can you rebase this PR?

@compnerd, @tstellar, @petrhosek can we make another review iteration of this PR?

**Review by compnerd — COMMENTED** [review]
The non-CMake portions seem fine. I'd say get @petrhosek's signoff for the CMake bits.

File: llvm/include/llvm/Support/Compiler.h
Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h:136]: I don't know it was existing code that got rearranged slightly, it was controlling LLVM_EXTERNAL_VISIBILITY before

File: llvm/include/llvm/Support/Compiler.h
Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h:138]: What is the alternative for non windows platforms other than this? I used LLVM_ALWAYS_EXPORT for some special JIT debugger registration functions that were annotated with LLVM_ATTRIBUTE_VISIBILITY_DEFAULT before.

**Comment by fsfod** [comment]
Ping @petrhosek can you check my changes.

File: llvm/include/llvm/Support/Compiler.h
Comment (compnerd) [reply llvm/include/llvm/Support/Compiler.h:138]: Explicitly spelling out the attribute OSS what I was suggesting. On Linux, protected visibility has interesting behavior where symbols are not interpositionable and are externally visible. Using that is potentially useful for load times on Linux for exported symbols.

Comment (fsfod) [reply llvm/include/llvm/Support/Compiler.h:138]: I didn't know about protected, wouldn't it also indirectly stop symbols being merging across shared libraries and executables by the loader? maybe that would be what you always want for the JIT debugger symbols.

File: llvm/CMakeLists.txt
Comment (petrhosek) [inline llvm/CMakeLists.txt:854]: We usually use `ON` and `OFF`.
```suggestion
set(CAN_BUILD_LLVM_DYLIB OFF)
if(NOT MSVC OR LLVM_BUILD_LLVM_DYLIB_VIS)
  set(CAN_BUILD_LLVM_DYLIB ON)
endif()
```

File: llvm/CMakeLists.txt
Comment (petrhosek) [inline llvm/CMakeLists.txt:869]: We prefer lowercase spelling for CMake function.
```suggestion
cmake_dependent_option(LLVM_LINK_LLVM_DYLIB "Link tools against the libllvm dynamic library" OFF
                       "CAN_BUILD_LLVM_DYLIB" OFF)

set(LLVM_BUILD_LLVM_DYLIB_default OFF)
if(LLVM_LINK_LLVM_DYLIB OR LLVM_BUILD_LLVM_C_DYLIB)
  set(LLVM_BUILD_LLVM_DYLIB_default ON)
endif()
cmake_dependent_option(LLVM_BUILD_LLVM_DYLIB "Build libllvm dynamic library" ${LLVM_BUILD_LLVM_DYLIB_default}
                       "CAN_BUILD_LLVM_DYLIB" OFF)

cmake_dependent_option(LLVM_DYLIB_EXPORT_INLINES "Force inline members of classes to be DLL exported when
                       building with clang-cl so the libllvm DLL is compatible with MSVC"
                       OFF
                       "MSVC;LLVM_BUILD_LLVM_DYLIB_VIS" OFF)
```

File: llvm/CMakeLists.txt
Comment (petrhosek) [inline llvm/CMakeLists.txt:852]: Where is `LLVM_BUILD_LLVM_DYLIB_VIS` defined? What does the `_VIS` suffix stand for?

Comment (fsfod) [reply llvm/CMakeLists.txt:852]: Its a temporary option to allow testing the build with explicit visibility on Windows and Linux the idea is it will go away in the end, when I've gotten all my changes merged. It has to be separate from LLVM_BUILD_LLVM_DYLIB because there will be linker errors from symbols not exported when building llvm tools until the whole llvm codebase has export visibility macros added to it.
The idea was it was not a publicly defined option so people don't try use and submit bug reports against it. If you want I can publicly define it with a don't use for testing purposes only description.

**Comment by fsfod** [comment]
Ping @petrhosek

**Comment by mstorsjo** [comment]
This patch breaks mingw builds.

Currently, it's very possible to build for mingw targets with `-DLLVM_LINK_LLVM_DYLIB=ON`; mingw toolchains default to exporting all symbols if there are no ones marked dllexport, and ELF-style attributes like `__attribute__((visibility("hidden")))` and options like `-fvisibility=hidden` works for excluding individual symbols or translation units from the automatic export.

I don't mind switching to the new approach of explicit dllexport/import of things, once it is working and complete, but I don't want to regress building with `-DLLVM_LINK_LLVM_DYLIB=ON` in the meantime.

With a patch like this, building for mingw targets keeps working like it used to:
```diff
--- a/llvm/include/llvm/Support/Compiler.h
+++ b/llvm/include/llvm/Support/Compiler.h
@@ -174,7 +174,7 @@
 #define LLVM_ABI
 #define LLVM_TEMPLATE_ABI
 #define LLVM_EXPORT_TEMPLATE
-#elif defined(_WIN32)
+#elif defined(_WIN32) && !defined(__MINGW32__)
 #if defined(LLVM_EXPORTS)
 #define LLVM_ABI __declspec(dllexport)
 #define LLVM_TEMPLATE_ABI
@@ -184,7 +184,7 @@
 #define LLVM_TEMPLATE_ABI __declspec(dllimport)
 #define LLVM_EXPORT_TEMPLATE
 #endif
-#elif defined(__ELF__)
+#elif defined(__ELF__) || defined(__MINGW32__)
 #define LLVM_ABI LLVM_ATTRIBUTE_VISIBILITY_DEFAULT
 #define LLVM_TEMPLATE_ABI LLVM_ATTRIBUTE_VISIBILITY_DEFAULT
 #define LLVM_EXPORT_TEMPLATE
```

If you want to test that this keeps working, I tried setting up a simple build file for GitHub Actions: https://github.com/mstorsjo/llvm-project/commit/gha-build-windows If you push a branch with this commit, to your own fork, it should test building with MSVC, clang-cl and llvm-mingw (with and without dylibs enabled).

**Comment by fsfod** [comment]
Thanks for catching that, it wasn't my intention for any targets to be non functional until all the required changes have gone in, after which i can change default symbol visibility set to hidden for elf targets.
Something i do wonder about is if loadable llvm and clang plugins work on mingw currently? some of my later changes to make them work for windows could also fix them for mingw if they don't currently work.

**Comment by mstorsjo** [comment]
> Something i do wonder about is if loadable llvm and clang plugins work on mingw currently? some of my later changes to make them work for windows could also fix them for mingw if they don't currently work.

I’m not entirely sure. Iirc there are multiple kinds of plugins within all of LLVM, some kinds do work, while others don’t, if I remember correctly.

**Comment by fsfod** [comment]
Ping @petrhosek

File: llvm/CMakeLists.txt
Comment (petrhosek) [reply llvm/CMakeLists.txt:852]: You could consider making it an internal variable:
```
set(LLVM_BUILD_LLVM_DYLIB_VIS OFF CACHE INTERNAL "")
```

Comment (fsfod) [reply llvm/CMakeLists.txt:856]: I didn't realize it had that mode, I've update code to define it like that.

File: llvm/tools/llvm-shlib/CMakeLists.txt
Comment (petrhosek) [reply llvm/tools/llvm-shlib/CMakeLists.txt:55]: Not yet, that feature was introduced in CMake 3.24 but the minimum version currently supported by LLVM is 3.20.

**Review by petrhosek — APPROVED** [review]

**Comment by compnerd** [comment]
I think that this is ready to go at this point. This unblocks the subsequent work to start marking the boundaries for the library ABI.

**Comment by fsfod** [comment]
I'm not sure if LLVM_LIBRARY_VISIBILITY error on gvn is the the issue @tstellar mentioned about some attribute types are only supported in different places by some compilers and older versions of clang.

**Comment by tstellar** [comment]
@fsfod Is there a failing buildbot with a log I can look at?

**Comment by fsfod** [comment]
https://lab.llvm.org/buildbot/#/builders/12/builds/5294
https://lab.llvm.org/buildbot/#/builders/78/builds/5463

**Comment by fsfod** [comment]
The error seems to imply LLVM_LIBRARY_VISIBILITY is undefined but compiler.h is included by the file

**Comment by tstellar** [comment]
I thought it might be this: https://github.com/tstellar/template-visibility/blob/testing/templates.h#L37

That buildbot is using clang 14.

**Comment by tstellar** [comment]
> I thought it might be this: https://github.com/tstellar/template-visibility/blob/testing/templates.h#L37
> 
> That buildbot is using clang 14.

If this is indeed the problem, i would recommend just disabling the annotations for clang < 15. (i.e. have the macros expand to nothing).

**Comment by fsfod** [comment]
Would changing the preprocessor condition from 
``` c++
#if LLVM_HAS_CPP_ATTRIBUTE(gnu::visibility) && defined(__GNUC__) &&            \
    !defined(__clang__)
```
to 
``` c++
#if LLVM_HAS_CPP_ATTRIBUTE(gnu::visibility) && defined(__GNUC__) &&            \
    (!defined(__clang__) || _clang_major__ < 15)
``` 

not be enough to fix it?

**Comment by tstellar** [comment]
@fsfod possibly, I'm not sure what implications that would have in othe parts of the code.

**Pull request merged**
